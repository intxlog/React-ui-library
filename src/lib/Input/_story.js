import React from 'react'
import { storiesOf } from '@storybook/react';
import Input from './'

storiesOf('Input', module)
.add('default', () => (
  <Input></Input>
))
.add('default with placeholder', () => (
  <Input
    placeholder={`Example Placeholder`}
  ></Input>
))
.add('default (required)', () => (
  <Input
    required
  ></Input>
))
.add('default (disabled)', () => (
  <Input disabled></Input>
))
.add('default with defaultValue (disabled)', () => (
  <Input disabled defaultValue={`Example Text`}></Input>
))
.add('default with infoText', () => (
  <Input infoText={`Example info text`}></Input>
))
.add('default with infoText and error', () => (
  <Input infoText={`Example info text`} error={true}></Input>
))
.add('email', () => (
  <Input
    type='email'
  ></Input>
))
.add('email with placeholder', () => (
  <Input
    type='email'
    placeholder={`Example Placeholder`}
  ></Input>
))
.add('email (required)', () => (
  <Input
    type='email'
    required
  ></Input>
))
.add('email (disabled)', () => (
  <Input disabled
    type='email'
  ></Input>
))
.add('email with defaultValue (disabled)', () => (
  <Input disabled
    type='email'
    defaultValue={'test@test.com'}
  ></Input>
))
.add('email with infoText', () => (
  <Input 
    type='email'
    infoText={`Enter email`}
    ></Input>
))
.add('email with infoText and error', () => (
  <Input 
    type='email'
    infoText={`email is incorrect`} 
    error={true}
    ></Input>
))
.add('email isValid event fires', () => (
  <Input
    type='email'
    required={true}
    defaultValue={'test@test.com'}
    isValid={(bool) => {console.log(bool)}}
  ></Input>
))
.add('password', () => (
  <Input
    type='password'
  ></Input>
))
.add('password with placeholder', () => (
  <Input
    type='password'
    placeholder={`Example Placeholder`}
  ></Input>
))
.add('password (required)', () => (
  <Input
    type='password'
    required
  ></Input>
))
.add('password (Validation) Must Use One CAPITAL letter', () => (
  <Input
    type='password'
    placeholder={`One CAPITAL Letter to Pass`}
    verify   
  ></Input>
))
.add('password (disabled)', () => (
  <Input disabled
    type='password'
  ></Input>
))
.add('password with defaultValue (disabled)', () => (
  <Input disabled
    type='password'
    defaultValue={'***********'}
  ></Input>
))
.add('password with infoText', () => (
  <Input 
    type='password'
    infoText={`Enter password`}
    ></Input>
))
.add('password with infoText and error', () => (
  <Input 
    type='password'
    infoText={`Password or Username is incorrect`} 
    error={true}
    ></Input>
))
.add('password isValid event fires', () =>(
  <Input
    type='password'
    required={true}
    defaultValue={'password123'}
    isValid={(bool) => {console.log(bool)}}
    ></Input>
))

storiesOf('TextArea', module)
.add('default', () => (
  <Input type={`textArea`}></Input>
))
.add('default with placeholder', () => (
  <Input type={`textArea`} placeholder={`Example of a placeholder`}></Input>
))
.add('default (required)', () => (
  <Input
    type={`textArea`}
    required
  ></Input>
))
.add('default (disabled)', () => (
  <Input type={`textArea`} disabled></Input>
))
.add('default with defaultValue (disabled)', () => (
  <Input 
    type={`textArea`} 
    disabled
    defaultValue={`Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse sollicitudin velit ac tellus posuere, vel rutrum dolor pellentesque. Duis dapibus sapien in mauris pretium volutpat. Vivamus id purus in erat pharetra ultrices eget sit amet orci. Ut ornare dolor sit amet rutrum euismod. Suspendisse potenti. Nunc sed urna quis augue dictum rhoncus. Vivamus vehicula sapien sed dolor gravida, at imperdiet mi aliquet. Curabitur venenatis gravida metus et tempor. Phasellus quis tortor lacinia, gravida magna sit amet, tincidunt odio. Donec nulla purus, interdum nec condimentum ut, rutrum quis arcu.`}
  ></Input>
))
.add('default with defaultValue', () => (
  <Input type={`textArea`} defaultValue={`Example of a default value`}></Input>
))
.add('default with infoText', () => (
  <Input 
    type='textArea'
    infoText={`Enter textArea`}
  ></Input>
))
.add('default with infoText and error', () => (
  <Input 
    type='textArea'
    infoText={`textArea is incorrect`} 
    error={true}
  ></Input>
))
.add('default isValid event fires', () =>(
  <Input
    type='textArea'
    required={true}
    defaultValue={'Default textArea value'}
    isValid={(bool) => {console.log(bool)}}
  ></Input>
))

storiesOf('Input Select', module)
.add('default', () => (
  <Input
    type='select'
  >
  <option value="default">Choose an option</option>
  </Input>
))
.add('default (disabled)', () => (
  <Input
    type='select'
    disabled
  >
  </Input>
))
.add('default with options (disabled)', () => (
  <Input
    type='select'
    defaultValue='default'
    disabled
  >
    <option value="default" disabled>Choose an option</option>
    <option>Option 1</option>
    <option>Option 2</option>
    <option>Option 3</option>
  </Input>
))
.add('with options', () => (
  <Input
    type='select'
    defaultValue='default'
  >
    <option value="default" disabled>Choose an option</option>
    <option>Option 1</option>
    <option>Option 2</option>
    <option>Option 3</option>
  </Input>
))
.add('default with infoText', () => (
  <Input 
    type='select'
    infoText={`Default info text`}
    >
    <option value="default" disabled>Choose an option</option>
    <option>Option 1</option>
    <option>Option 2</option>
    <option>Option 3</option>
  </Input>
))
.add('default with infoText and error', () => (
  <Input 
    type='select'
    infoText={`option is incorrect`}  
    error={true}    
    >
    <option value="default" disabled>Choose an option</option>
    <option>Option 1</option>
    <option>Option 2</option>
    <option>Option 3</option>
  </Input>
))
.add('default isValid event fires', () =>(
  <Input
    type='select'
    required={true}
    defaultValue={'Default default value'}
    isValid={(bool) => {console.log(bool)}}
    >
    <option value="default" disabled>Choose an option</option>
    <option>Option 1</option>
    <option>Option 2</option>
    <option>Option 3</option>
  </Input>
))


storiesOf('Input Radio', module)
.add('default', () => (
  <>
    <Input
      type='radio'
      name={'test'}
      value={`option1`}
      idForLabel={`option1`}
      labelText={'Option 1'}
    ></Input>
    <Input
      type='radio'
      name={'test'}
      value={`option2`}
      idForLabel={`option2`}
      labelText={'Option 2'}
    ></Input>
  </>
))
.add('defaultChecked', () => (
  <>
    <Input
      type='radio'
      name={'test'}
      value={`option1`}
      idForLabel={`option1`}
      labelText={'Option 1'}
      defaultChecked
    ></Input>
    <Input
      type='radio'
      name={'test'}
      value={`option2`}
      idForLabel={`option2`}
      labelText={'Option 2'}
    ></Input>
  </>
))
.add('default (disabled)', () => (
  <>
    <Input
      type='radio'
      name={'test'}
      value={`option1`}
      idForLabel={`option1`}
      labelText={'Option 1'}
      disabled
    ></Input>
    <Input
      type='radio'
      name={'test'}
      value={`option2`}
      idForLabel={`option2`}
      labelText={'Option 2'}
      disabled
    ></Input>
  </>
))
.add('default (disabled) with defaultChecked', () => (
  <>
    <Input
      type='radio'
      name={'test'}
      value={`option1`}
      idForLabel={`option1`}
      labelText={'Option 1'}
      defaultChecked
      disabled
    ></Input>
    <Input
      type='radio'
      name={'test'}
      value={`option2`}
      idForLabel={`option2`}
      labelText={'Option 2'}
      disabled
    ></Input>
  </>
))
.add('default with infoText', () => (
  <>
    <Input
      type='radio'
      name={'test'}
      value={`option1`}
      idForLabel={`radio1`}
      labelText={'Option 1'}
      infoText={`Option 1 radio`} 
      defaultChecked
    ></Input>
    <Input
      type='radio'
      name={'test'}
      value={`option2`}
      idForLabel={`radio2`}
      labelText={'Option 2'}
      infoText={`Option 2 radio`} 
    ></Input>
  </>
))
.add('default with infoText and error', () => (
  <>
    <Input
      type='radio'
      name={'test'}
      value={`option1`}
      idForLabel={`radio1`}
      labelText={'Option 1'}
      infoText={`Option 1 is incorrect`} 
      error={true}
      defaultChecked
    ></Input>
    <Input
      type='radio'
      name={'test'}
      value={`option2`}
      idForLabel={`radio2`}
      labelText={'Option 2'}
      infoText={`Option 2 is incorrect`} 
      error={true}
    ></Input>
  </>
))

storiesOf('Input Checkbox', module)
.add('default', () => (
  <>
    <Input
      type='checkbox'
      value={`Checkbox`}
      idForLabel={`checkbox1`}
      labelText={'1st Awesome Checkbox'}
    ></Input>
    <Input
      type='checkbox' 
      value={`Checkbox`}
      idForLabel={`checkbox2`}
      labelText={'2nd Awesome Checkbox'}
    ></Input>
  </>
))
.add('defaultChecked', () => (
  <>
    <Input
      type='checkbox'
      value={`Checkbox`}
      idForLabel={`checkbox1`}
      labelText={'1st Awesome Checkbox'}
      defaultChecked
    ></Input>
    <Input
      type='checkbox' 
      value={`Checkbox`}
      idForLabel={`checkbox2`}
      labelText={'2nd Awesome Checkbox'}
    ></Input>
  </>
))
.add('default (disabled)', () => (
  <>
    <Input
      type='checkbox'
      value={`Checkbox`}
      idForLabel={`checkbox1`}
      labelText={'1st Awesome Checkbox'}
      disabled
    ></Input>
    <Input
      type='checkbox' 
      value={`Checkbox`}
      idForLabel={`checkbox2`}
      labelText={'2nd Awesome Checkbox'}
      disabled
    ></Input>
  </>
))
.add('defaultChecked (disabled)', () => (
  <>
    <Input
      type='checkbox'
      value={`Checkbox`}
      idForLabel={`checkbox1`}
      labelText={'1st Awesome Checkbox'}
      defaultChecked
      disabled
    ></Input>
    <Input
      type='checkbox' 
      value={`Checkbox`}
      idForLabel={`checkbox2`}
      labelText={'2nd Awesome Checkbox'}
      disabled
    ></Input>
  </>
))
.add('default with infoText', () => (
  <>
    <Input
      type='checkbox'
      value={`Checkbox`}
      idForLabel={`checkbox1`}
      labelText={'1st Awesome Checkbox'}
      infoText={`Checkbox 1`} 
    ></Input>
    <Input
      type='checkbox' 
      value={`Checkbox`}
      idForLabel={`checkbox2`}
      labelText={'2nd Awesome Checkbox'}
      infoText={`Checkbox 2`} 
    ></Input>
  </>
))
.add('default with infoText and error ', () => (
  <>
    <Input
      type='checkbox'
      value={`Checkbox`}
      idForLabel={`checkbox1`}
      labelText={'1st Awesome Checkbox'}
      infoText={`Option 1 is incorrect`} 
      error={true}
    ></Input>
    <Input
      type='checkbox' 
      value={`Checkbox`}
      idForLabel={`checkbox2`}
      labelText={'2nd Awesome Checkbox'}
      infoText={`Option 2 is incorrect`} 
      error={true}
    ></Input>
  </>
))