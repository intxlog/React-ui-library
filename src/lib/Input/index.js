import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

//import css
import styles from './styles.module.scss'

//import validators
import validateString from './../../validators'
import validateRequired from './../../validators/required'

//import components
import TextInput from '../TextInput'
import TextArea from '../TextArea'
import Checkbox from '../Checkbox'
import RadioButton from '../RadioButton'
import Select from '../Select'

// logic behind all the different types of inputs
class Input extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      value: '',
      entered: false,
      isValid: false,
      validValue: null,
      error: false,
      infoText: null
    }
  }

  componentDidMount() {
    //fire the isValid function letting the user know the input is not valid by default
    //check to make sure there is not a defaultValue
    if (this.props.defaultValue) {
      let val = this.props.defaultValue
      
      //Clear out the value if the value is IUILIBdefault and make it nothing. This is value that just makes it so we make default values in selects
      if (val === `IUILIBdefault`) {
        val = ``

        //fire the isValid function if one is passed in
        this.props.isValid(this.state.isValid)
      } else {
        //set state as the defaultValue
        this.setState({
          value: val
        })

        //validate the value
        this.validate(val)
      }
    } else { //there is not a defaultValue
      //fire the isValid function if one is passed in
      this.props.isValid(this.state.isValid)
    } //end defaultValue check
  }

  componentDidUpdate(prevProps, prevState) {
    //fire isValid function from props if isValid from state changed
    if (this.state.isValid !== prevState.isValid) {
      this.props.isValid(this.state.isValid)
    }

    //check to see if the formSubmitted prop becomes true and validate the field
    if (this.props.formSubmitted !== prevProps.formSubmitted && this.props.formSubmitted) {
      //change the entered value in state to true to mimick the user entering the field
      this.setState({
        entered: true
      })

      //validate the value
      this.validate(this.state.value)
    }
  }

  handleOnChange = (e) => {
    const val = e.target.value

    //check if the user entered the input once before
    if (this.state.entered) {
      //run validation
      this.validate(val)
    }
    
    //update the state and fire the onChange function if one was passed in
    this.setState({value: val})
    this.props.onChange(val)  
  }

  handleOnBlur = () => {
    const val = this.state.value

    //set entered to true
    this.setState({entered: true})

    //run validation
    this.validate(val)

    //fire the blur event that was passed in 
    this.props.onBlur(val)
  }

  checkRequired = (val) => {
    const res = validateRequired(val)
    const isValid = res.valid
    const message = res.message
    const newValue = res.value

    if (!isValid) {
      this.setState({
        error: true,
        infoText: message
      })
    } else {
      this.setState({
        error: false,
        infoText: null,
        validValue: newValue
      })
    }

    //return a bool value
    return isValid
  }

  checkValid = (val) => {
    //shorthand if statement
    const res = this.props.customValidationFunc ? this.props.customValidationFunc(val) : validateString(val, this.props.type)
    const isValid = res.valid
    const message = res.message
    const newValue = res.value

    //if it is not valid then set the error state
    if (!isValid) {
      this.setState({
        error: true,
        infoText: message
      })
    } else {
      this.setState({
        error: false,
        infoText: null,
        validValue: newValue
      })
    }

    //return a bool value
    return isValid
  }

  validate = (val) => {
    //create a flag to see if we should fire the onValid function
    let isValid = false
    //create a flag to see if required has passed to see if we will need to validate
    let requiredPassed = false

    //check if this is required and passes the requirements
    if (this.props.required) {
      isValid = this.checkRequired(val)
      requiredPassed = this.checkRequired(val)
    } else {
      requiredPassed = true
    }

    //check if validate prop is true and it passes the requirements
    if (this.props.validate && requiredPassed) {
      isValid = this.checkValid(val)
    }

    //set isValid in state to the proper value
    this.setState({
      isValid
    })
  }

  render(){
    //var to handle the html type
    let type = null
    //var to store the html element we are going to display
    let element = null
    
    //logic for the different html input types
    switch (this.props.type) {
      case `password`:
        type = `password`
        break
      default:
        type = `text`
        break
    }

    //switch statement to handle different html cases
    switch (this.props.type) {
      case `select`:
        element = <Select
          id={this.props.idForLabel}
          defaultValue={this.props.defaultValue}
          disabled={this.props.disabled}
          onChange={this.handleOnChange}
          onBlur={this.handleOnBlur}
          error={this.props.error}
        >
          {this.props.children}
        </Select>
        break

      case `radio`:
        element = <RadioButton
          id={this.props.idForLabel}
          value={this.props.value}
          name={this.props.name}
          labelText={this.props.labelText}
          defaultChecked={this.props.defaultChecked}
          disabled={this.props.disabled}
          onChange={this.handleOnChange}
          onBlur={this.handleOnBlur}
        />
        break

      case `checkbox`:
        element = <Checkbox
          id={this.props.idForLabel}
          value={this.props.value}
          name={this.props.name}
          defaultChecked={this.props.defaultChecked}
          onChange={this.handleOnChange}
          onBlur={this.handleOnBlur}
          labelText={this.props.labelText}
        ></Checkbox>
        break

      case `textArea`:
        element = <TextArea
          id={this.props.idForLabel}
          inlineStyles={this.props.inlineStyles}
          placeholder={this.props.placeholder}
          disabled={this.props.disabled}
          defaultValue={this.props.defaultValue}
          onChange={this.handleOnChange}
          onBlur={this.handleOnBlur}
          error={this.props.error}
        ></TextArea>
        break

      default:
        element = <TextInput
          id={this.props.idForLabel}
          inlineStyles={this.props.inlineStyles}
          placeholder={this.props.placeholder}
          disabled={this.props.disabled}
          defaultValue={this.props.defaultValue}
          type={type}
          onChange={this.handleOnChange}
          onBlur={this.handleOnBlur}
        ></TextInput>
        break
    }
    
    return (
      <>
        {element}
        <p className={classNames({
          [styles.infoText]: true,
          [styles.error]: this.props.error || this.state.error
        })}>{this.props.infoText || this.state.infoText}</p>
      </>
    )
  }
}

//rules for props being passed in
Input.propTypes = {
  idForLabel: PropTypes.string,
  labelText: PropTypes.string,
  inlineStyles: PropTypes.object,
  placeholder: PropTypes.string,
  disabled: PropTypes.bool,
  infoText: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool
    ]),
  name: PropTypes.string,
  error: PropTypes.bool,
  defaultValue: PropTypes.string,
  defaultChecked: PropTypes.bool,
  validate: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.oneOf(['text', 'email', 'password', 'select', 'radio', 'checkbox', 'textArea']).isRequired,
  customValidationFunc: PropTypes.func,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  isValid: PropTypes.func,
  formSubmitted: PropTypes.bool
}

// Specifies the default values for props:
Input.defaultProps = {
  labelText: `None Provided`,
  placeholder: ``,
  infoText: ``,
  idForLabel: ``,
  inlineStyles: {},
  error: false,
  defaultValue: '',
  validate: true,
  required: false,
  type: 'text',
  onChange: () => {},
  onBlur: () => {},
  isValid: () => {}
}

//maintain the name for documentation purposes
Input.displayName = 'Input'

export default Input