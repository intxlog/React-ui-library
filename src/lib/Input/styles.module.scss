.infoText {
  display: block;
  font-size: 12px;
  line-height: 16px;
  margin: 0;
  padding-left: 6px;
  margin-bottom: 4px;
  //keep the elements space even when it is blank
  min-height: 16px;

  &.error {
    color: red;
  }
}

.errorBox {
  min-height: 16px;
}