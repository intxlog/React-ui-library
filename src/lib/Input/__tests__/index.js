import React from 'react'
import { shallow } from 'enzyme'
import Input from '..'

describe('Input', () => {
  //mock functions
  const onChangeMock = jest.fn()
  const onBlurMock = jest.fn()
  const isValidMock = jest.fn()

  //Spies
  let checkRequiredSpy
  let validateSpy
  
  let props = {
    onChange: onChangeMock,
    onBlur: onBlurMock,
    isValid: isValidMock
  }
  
  //create a shallow wrapper that can be reused
  let wrapper
  //initialize the components
  beforeEach(() => {
    wrapper = shallow(<Input {...props}/>)
    checkRequiredSpy = jest.spyOn(wrapper.instance(), 'checkRequired')
    validateSpy = jest.spyOn(wrapper.instance(), 'validate')
  })


  //reset the mocks after each test
  afterEach(() => {
    onChangeMock.mockReset()
    onBlurMock.mockReset()
    isValidMock.mockReset()
    checkRequiredSpy.mockRestore()
    validateSpy.mockRestore()
  })

  it('renders correctly', () => {
    // const wrapper = shallow( <Input/> )
    expect(wrapper).toMatchSnapshot()
  })
  describe('when disabled is passed in', () => {
    it('renders correctly', () => {
      wrapper.setProps({
          disabled: true
      })
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when validate is passed in', () => {
    it('renders correctly', () => {
      wrapper.setProps({
        validate: true
      })
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when required is passed in', () => {
    it('renders correctly', () => {
      wrapper.setProps({
        required: false
      })
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when type is password', () => {
    it('renders correctly', () => {
      wrapper.setProps({
        type:'password',
        password: true
      })
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when idForLabel prop is passed in', () => {
    it('renders correctly', () => {
      wrapper.setProps({
        idForLabel: `exampleID`
      })
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when placeholder prop is passed in', () => {
    it('renders correctly', () => {
      wrapper.setProps({
        placeholder: `Example Text`
      })
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when infoText prop is passed in', () => {
    it('renders correctly', () => {
      wrapper.setProps({
        infoText: `Example Text`
      })
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when error prop is true', () => {
    it('renders correctly', () => {
      wrapper.setProps({
        error: true
      })
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when value prop is passed in', () => {
    it('renders correctly', () => {
      wrapper.setProps({
        value: `Example Text`
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  //checkbox testing
  describe('checkbox', () => {
    it('renders correctly', () => {
      wrapper.setProps({
        type:'checkbox',
        idForLabel:`exampleID`
      })
      expect(wrapper).toMatchSnapshot()
    })
  })


  //Radio testing
  describe('radio', () => {
    it('renders correctly', () => {
      wrapper.setProps({
        type:'radio',
        idForLabel:`exampleID`,
        value: true
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  //Select testing
  describe('select', () => {
    describe('renders correctly', () => {
      it('when type is select', () => {
        wrapper.setProps({
          type: 'select'
        })
        expect(wrapper).toMatchSnapshot()
      }),
      it('when disabled is true', () => {
        wrapper.setProps({
          type: 'select',
          disabled: true
        })
        expect(wrapper).toMatchSnapshot()
      })
    })
  })

  //TextArea testing
  describe('textArea', () => {
    it('renders correctly', () => {
      wrapper.setProps({
        type: 'textArea'
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('textArea', () => {
    describe('renders correctly', () => {
      it('when type is textArea', () => {
        wrapper.setProps({
          type: 'textArea'
        })
        expect(wrapper).toMatchSnapshot()
      }),
      it('when disabled is true', () => {
        wrapper.setProps({
          type: 'textArea',
          disabled: true
        })
        expect(wrapper).toMatchSnapshot()
      })
    })
  })

  //Internal Methods
  describe('when checkRequired is called', () => {
    describe('with a valid value', () => {
      it('returns the correct value', () => {
        expect(wrapper.instance().checkRequired('foo')).toEqual(true)
      })
    })
    describe('with a invalid value', () => {
      it('returns the correct value', () => {
        expect(wrapper.instance().checkRequired('')).toEqual(false)
      })
    })
  })

  describe('when checkValid is called', () => {
    describe('with ""', () => {
      it('returns the correct value', () => {
        wrapper.setProps({
          type: `email`
        })
        expect(wrapper.instance().checkValid('')).toEqual(false)
      })
    })
    describe('with test@test.com', () => {
      it('returns the correct value', () => {
        wrapper.setProps({
          type: `email`
        })
        expect(wrapper.instance().checkValid('test@test.com')).toEqual(true)
      })
    })
  })
  
  describe('when validate is called', () => {
    describe('with a value and required prop is true', () => {
      it('calls the correct method', () => {
        wrapper.setProps({
          required: true
        })

        //call the method
        wrapper.instance().validate('test')

        //check the value of the spy
        expect(checkRequiredSpy).toHaveBeenCalled()
      })
    })
    describe('with a value and disabled prop is true', () => {
      it('calls the correct method', () => {
        wrapper.setProps({
          disabled: true
        })

        //call the method
        wrapper.instance().validate('test')

        expect(wrapper.state().isValid).toEqual(true)
      })
    })
  })

  describe('when handleOnChange is called', () => {
    describe('when the entered value in state is true', () => {
      it('calls the validate method', () => {
        const e = {
          target: {
            value: 'test'
          }
        }

        //set state
        wrapper.setState({
          entered: true
        })

        //call the method
        wrapper.instance().handleOnChange(e)
  
        expect(validateSpy).toHaveBeenCalled()
      })  
    })
  })

  describe('when handleOnBlur is called', () => {
    it('calls the correct function', () => {
      wrapper.setState({
        value: 'test'
      })
      wrapper.instance().handleOnBlur()
      expect(onBlurMock).toHaveBeenCalled()
      expect(onBlurMock).toHaveBeenCalledWith('test')
    })
  })

  describe('when component mounts', () => {
    describe('when there is no default value', () => {
      it('calls the correct function', () => {
        expect(isValidMock).toHaveBeenCalled()
        expect(isValidMock).toHaveBeenCalledWith(false)
      })
    })
    describe('when given a defaultValue of "test"', () => {
      it('calls the correct function and sets the correct state', () => {
        wrapper.setProps({
          defaultValue: 'test'
        })

        //call the method
        wrapper.instance().componentDidMount()
        expect(wrapper.state().value).toEqual('test')
        expect(validateSpy).toHaveBeenCalled()
        expect(validateSpy).toHaveBeenCalledWith('test')
      })
    })  
    describe('when given a defaultValue of "IUILIBdefault"', () => {
      it('calls the correct function', () => {
        wrapper.setProps({
          defaultValue: 'IUILIBdefault'
        })

        //call the method
        wrapper.instance().componentDidMount()
        expect(isValidMock).toHaveBeenCalled()
        expect(isValidMock).toHaveBeenCalledWith(false)
      })
    }) 
  })

  describe('when componentDidUpdate is called', () => {
    describe('formSubmitted prop changes', () => {
      it('calls the correct function and state is correct', () => {
        //set the props
        wrapper.setProps({
          formSubmitted: 0
        })

        //change the props to trigger componentDidUpdate
        wrapper.setProps({
          formSubmitted: 1
        })

        expect(wrapper.state().entered).toEqual(true)
        expect(validateSpy).toHaveBeenCalled()
        expect(validateSpy).toHaveBeenCalledWith('')
      })
    })
  })

  describe('isValid state changes', () => {
    it('calls the correct function', () => {
      //change the state to trigger componentDidUpdate
      wrapper.setState({
        isValid: true
      })

      expect(isValidMock).toHaveBeenCalled()
      expect(isValidMock).toHaveBeenCalledWith(true)
    })
  })
})
