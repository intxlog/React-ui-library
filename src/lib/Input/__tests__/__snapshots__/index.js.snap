// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input checkbox renders correctly 1`] = `
<Fragment>
  <Checkbox
    id="exampleID"
    labelText="None Provided"
    onBlur={[Function]}
    onChange={[Function]}
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input radio renders correctly 1`] = `
<Fragment>
  <RadioButton
    id="exampleID"
    labelText="None Provided"
    onBlur={[Function]}
    onChange={[Function]}
    value={true}
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input renders correctly 1`] = `
<Fragment>
  <TextInput
    defaultValue=""
    error={false}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input select renders correctly when disabled is true 1`] = `
<Fragment>
  <Select
    defaultValue=""
    disabled={true}
    error={false}
    id=""
    onBlur={[Function]}
    onChange={[Function]}
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input select renders correctly when type is select 1`] = `
<Fragment>
  <Select
    defaultValue=""
    error={false}
    id=""
    onBlur={[Function]}
    onChange={[Function]}
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input textArea renders correctly 1`] = `
<Fragment>
  <TextArea
    defaultValue=""
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input textArea renders correctly when disabled is true 1`] = `
<Fragment>
  <TextArea
    defaultValue=""
    disabled={true}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input textArea renders correctly when type is textArea 1`] = `
<Fragment>
  <TextArea
    defaultValue=""
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input when disabled is passed in renders correctly 1`] = `
<Fragment>
  <TextInput
    defaultValue=""
    disabled={true}
    error={false}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input when error prop is true renders correctly 1`] = `
<Fragment>
  <TextInput
    defaultValue=""
    error={true}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input when idForLabel prop is passed in renders correctly 1`] = `
<Fragment>
  <TextInput
    defaultValue=""
    error={false}
    id="exampleID"
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input when infoText prop is passed in renders correctly 1`] = `
<Fragment>
  <TextInput
    defaultValue=""
    error={false}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input when placeholder prop is passed in renders correctly 1`] = `
<Fragment>
  <TextInput
    defaultValue=""
    error={false}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Example Text"
    type="text"
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input when required is passed in renders correctly 1`] = `
<Fragment>
  <TextInput
    defaultValue=""
    error={false}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input when type is password renders correctly 1`] = `
<Fragment>
  <TextInput
    defaultValue=""
    error={false}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="password"
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input when validate is passed in renders correctly 1`] = `
<Fragment>
  <TextInput
    defaultValue=""
    error={false}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <div
    className="errorBox"
  />
</Fragment>
`;

exports[`Input when value prop is passed in renders correctly 1`] = `
<Fragment>
  <TextInput
    defaultValue=""
    error={false}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <div
    className="errorBox"
  />
</Fragment>
`;
